src/scheduler.rs

```diff
@@ -0,0 +1,663 @@
+//! An unbounded set of futures.
+
+use Async;
+use executor::{self, UnsafeNotify, NotifyHandle};
+use task_impl::{AtomicTask, ThreadNotify};
+
+use std::cell::UnsafeCell;
+use std::fmt::{self, Debug};
+use std::marker::PhantomData;
+use std::mem;
+use std::ptr;
+use std::sync::atomic::Ordering::{Relaxed, SeqCst, Acquire, Release, AcqRel};
+use std::sync::atomic::{AtomicPtr, AtomicBool};
+use std::sync::{Arc, Weak};
+use std::usize;
+
+/// A generic task-aware scheduler.
+///
+/// This is used both by `FuturesUnordered` and the current-thread executor.
+pub struct Scheduler<T, W> {
+    inner: Arc<Inner<T, W>>,
+    len: usize,
+    head_all: *const Node<T, W>,
+}
+
+pub struct Notify<'a, T: 'a, W: 'a>(&'a Arc<Node<T, W>>);
+
+/// Wakeup a sleeper
```

Comment:
Probably worth clarifying what a "sleeper" is. My understanding is its something that is in charge of polling the scheduler (either a Task in the FuturesUnordered case or a thread in the current-thread executor).